File: speckit-chat-plugin/src/main/kotlin/com/speckit/plugin/ui/pipeline/StepDetailPanel.kt
package com.speckit.plugin.ui.pipeline

import com.intellij.openapi.project.Project
import com.intellij.ui.JBColor
import com.intellij.ui.RoundedLineBorder
import com.intellij.ui.components.JBScrollPane
import com.intellij.ui.components.JBTextArea
import com.speckit.plugin.model.ChatRun
import com.speckit.plugin.model.ChatRunStatus
import com.speckit.plugin.model.FeatureEntry
import com.speckit.plugin.model.FeaturePaths
import com.speckit.plugin.model.PipelineStepDef
import com.speckit.plugin.model.PipelineStepState
import com.speckit.plugin.model.StepStatus
import com.speckit.plugin.service.ChatRunLauncher
import com.speckit.plugin.service.PipelineService
import com.speckit.plugin.persistence.SessionPersistenceManager
import com.speckit.plugin.ui.SessionPanel
import com.speckit.plugin.ui.TaskListPanel
import com.speckit.plugin.ui.component.PipelineUiHelpers
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Component
import java.awt.Dimension
import java.awt.FlowLayout
import java.awt.Font
import java.io.File
import javax.swing.BorderFactory
import javax.swing.BoxLayout
import javax.swing.JButton
import javax.swing.JLabel
import javax.swing.JPanel

/**
 * Renders the right-side detail view for a selected pipeline step or feature.
 *
 * Extracted from [com.speckit.plugin.ui.PipelinePanel] to keep each file
 * under ~400 lines for AI agent comprehension.
 */
class StepDetailPanel(
    private val project: Project,
    private val chatPanel: SessionPanel,
    private val persistenceManager: SessionPersistenceManager?,
    private val launcher: ChatRunLauncher?
) : JPanel(BorderLayout()) {

    // ── Callbacks wired by PipelinePanel ──────────────────────────────────

    var onRunStep: ((PipelineStepDef, String) -> Unit)? = null
    var onReplyToSession: ((String, String) -> Unit)? = null
    var displayNumber: ((PipelineStepDef) -> String) = { it.number.toString() }
    var loadArgs: ((String) -> String?) = { null }

    // ── Public API ───────────────────────────────────────────────────────────

    fun showStep(
        step: PipelineStepDef,
        state: PipelineStepState?,
        paths: FeaturePaths?,
        lastImplementRun: ChatRun?
    ) {
        removeAll()
        if (state == null) { revalidate(); repaint(); return }

        val content = contentPanel()

        // Header
        content.add(JLabel("Step ${displayNumber(step)}: ${step.name}").apply {
            font = font.deriveFont(Font.BOLD, font.size + 4f)
            alignmentX = Component.LEFT_ALIGNMENT
        })
        if (step.isOptional) {
            content.add(JLabel("(optional)").apply {
                foreground = JBColor.GRAY
                alignmentX = Component.LEFT_ALIGNMENT
            })
        }
        content.add(JLabel(step.description).apply {
            alignmentX = Component.LEFT_ALIGNMENT
            border = BorderFactory.createEmptyBorder(4, 0, 12, 0)
        })

        // Status
        content.add(JLabel("Status: ${PipelineUiHelpers.statusText(state.status)}").apply {
            foreground = PipelineUiHelpers.statusColor(state.status)
            font = font.deriveFont(Font.BOLD)
            alignmentX = Component.LEFT_ALIGNMENT
            border = BorderFactory.createEmptyBorder(0, 0, 12, 0)
        })

        // Clarification markers (clarify step only)
        if (step.id == "clarify" && paths != null) {
            addClarificationMarkers(content, paths)
        }

        // Prerequisites
        if (step.prerequisites.isNotEmpty()) {
            content.add(PipelineUiHelpers.sectionHeader("Prerequisites:"))
            for (result in state.prerequisiteResults) {
                content.add(PipelineUiHelpers.checkResultLabel(result, project))
            }
            content.add(PipelineUiHelpers.verticalSpacer(8))
        }

        // Outputs
        if (step.outputs.isNotEmpty()) {
            content.add(PipelineUiHelpers.sectionHeader("Outputs:"))
            for (result in state.outputResults) {
                content.add(PipelineUiHelpers.checkResultLabel(result, project))

                // For the checklist step, list individual files under the directory output
                if (step.id == "checklist" && result.artifact.isDirectory && result.exists && result.resolvedFile != null) {
                    addChecklistFiles(content, result.resolvedFile)
                }
            }
            content.add(PipelineUiHelpers.verticalSpacer(8))
        }

        // Checklists (implement step only)
        if (step.id == "implement" && paths != null) {
            addChecklists(content, paths)
        }

        // Task list (tasks, implement, taskstoissues steps)
        if (step.id in setOf("tasks", "implement", "taskstoissues") && paths != null) {
            val taskListPanel = TaskListPanel(
                project, chatPanel, persistenceManager,
                enableActions = step.id == "implement",
                launcher = launcher
            )
            taskListPanel.update(paths.featureDir)
            taskListPanel.alignmentX = Component.LEFT_ALIGNMENT
            content.add(taskListPanel)
            content.add(PipelineUiHelpers.verticalSpacer(8))
        }

        // "Yes, proceed" reply button (implement step, running with session)
        val implRun = lastImplementRun
        if (step.id == "implement" && implRun != null
            && implRun.status == ChatRunStatus.RUNNING && implRun.sessionId != null) {

            val replyButton = JButton("Yes, proceed with implementation").apply {
                alignmentX = Component.LEFT_ALIGNMENT
                addActionListener { onReplyToSession?.invoke(implRun.sessionId!!, "yes, proceed with implementation") }
            }
            content.add(replyButton)
            content.add(PipelineUiHelpers.verticalSpacer(8))
        }

        // Hands off to
        if (step.handsOffTo.isNotEmpty()) {
            content.add(JLabel("Hands off to \u2192 ${step.handsOffTo.joinToString(", ")}").apply {
                foreground = JBColor.GRAY
                alignmentX = Component.LEFT_ALIGNMENT
                border = BorderFactory.createEmptyBorder(0, 0, 12, 0)
            })
        }

        // Arguments
        val isReadOnly = step.id == "specify"
        val argsField = JBTextArea(3, 0).apply {
            lineWrap = true
            wrapStyleWord = true
            margin = java.awt.Insets(6, 8, 6, 8)
            text = loadArgs(step.id) ?: step.defaultArgs
            isEditable = !isReadOnly
            if (isReadOnly) background = JBColor(Color(245, 245, 245), Color(60, 63, 65))
        }
        val argsHeader = JPanel(BorderLayout()).apply {
            isOpaque = false
            alignmentX = Component.LEFT_ALIGNMENT
            maximumSize = Dimension(Int.MAX_VALUE, 24)
            add(JLabel("Arguments:").apply {
                font = font.deriveFont(Font.BOLD)
            }, BorderLayout.WEST)
            if (!isReadOnly) {
                add(JButton("Default").apply {
                    isBorderPainted = false
                    isContentAreaFilled = false
                    foreground = JBColor.BLUE
                    cursor = java.awt.Cursor.getPredefinedCursor(java.awt.Cursor.HAND_CURSOR)
                    font = font.deriveFont(font.size - 1f)
                    toolTipText = "Reset to default"
                    addActionListener { argsField.text = step.defaultArgs }
                }, BorderLayout.EAST)
            }
            border = BorderFactory.createEmptyBorder(0, 0, 4, 0)
        }
        content.add(argsHeader)
        val argsScroll = JBScrollPane(argsField).apply {
            alignmentX = Component.LEFT_ALIGNMENT
            maximumSize = Dimension(Int.MAX_VALUE, 80)
            border = BorderFactory.createCompoundBorder(
                BorderFactory.createEmptyBorder(0, 0, 8, 0),
                RoundedLineBorder(JBColor.GRAY, 6)
            )
        }
        content.add(argsScroll)

        // Run button (hidden for Specify — use "Add New Specification" instead)
        if (!isReadOnly) {
            val runButton = JButton("Run ${step.name} \u25B7").apply {
                alignmentX = Component.LEFT_ALIGNMENT
                addActionListener { onRunStep?.invoke(step, argsField.text.trim()) }
            }
            content.add(runButton)
        }

        content.add(javax.swing.Box.createVerticalGlue())
        add(content, BorderLayout.CENTER)
        revalidate()
        repaint()
    }

    fun showFeature(
        entry: FeatureEntry,
        states: Map<PipelineStepDef, PipelineStepState>?,
        steps: List<PipelineStepDef>
    ) {
        removeAll()
        val content = contentPanel()

        content.add(JLabel(entry.dirName).apply {
            font = font.deriveFont(Font.BOLD, font.size + 4f)
            alignmentX = Component.LEFT_ALIGNMENT
        })
        content.add(JLabel("${entry.completedSteps}/${entry.totalOutputSteps} output steps completed").apply {
            alignmentX = Component.LEFT_ALIGNMENT
            border = BorderFactory.createEmptyBorder(4, 0, 12, 0)
        })

        if (states != null) {
            content.add(PipelineUiHelpers.sectionHeader("Steps:"))
            for (step in steps) {
                val status = states[step]?.status ?: StepStatus.NOT_STARTED
                content.add(JLabel("  ${PipelineUiHelpers.statusIcon(status)}  ${displayNumber(step)}. ${step.name} \u2014 ${PipelineUiHelpers.statusText(status)}").apply {
                    foreground = PipelineUiHelpers.statusColor(status)
                    alignmentX = Component.LEFT_ALIGNMENT
                })
            }
        }

        content.add(javax.swing.Box.createVerticalGlue())
        add(content, BorderLayout.CENTER)
        revalidate()
        repaint()
    }

    fun clear() {
        removeAll()
        revalidate()
        repaint()
    }

    // ── Private helpers ──────────────────────────────────────────────────────

    private fun contentPanel() = JPanel().apply {
        layout = BoxLayout(this, BoxLayout.Y_AXIS)
        border = BorderFactory.createEmptyBorder(12, 16, 12, 16)
    }

    private fun addClarificationMarkers(content: JPanel, paths: FeaturePaths) {
        val specFile = PipelineService.resolveArtifactFile("spec.md", false, paths.basePath, paths.featureDir)
        if (specFile != null && specFile.isFile) {
            val markerCount = Regex("\\[NEEDS CLARIFICATION").findAll(specFile.readText()).count()
            val markerLabel = if (markerCount == 0) {
                JLabel("No clarification markers in spec.md").apply {
                    foreground = PipelineUiHelpers.greenColor
                }
            } else {
                JLabel("$markerCount [NEEDS CLARIFICATION] marker(s) in spec.md").apply {
                    foreground = PipelineUiHelpers.orangeColor
                    cursor = java.awt.Cursor.getPredefinedCursor(java.awt.Cursor.HAND_CURSOR)
                    addMouseListener(object : java.awt.event.MouseAdapter() {
                        override fun mouseClicked(e: java.awt.event.MouseEvent) {
                            PipelineUiHelpers.openFileInEditor(project, specFile)
                        }
                    })
                    PipelineUiHelpers.addHoverEffect(this)
                }
            }
            markerLabel.alignmentX = Component.LEFT_ALIGNMENT
            markerLabel.border = BorderFactory.createEmptyBorder(0, 0, 12, 0)
            content.add(markerLabel)
        }
    }

    private fun addChecklistFiles(content: JPanel, directory: File) {
        val files = directory.listFiles()
            ?.filter { it.isFile && it.extension == "md" }
            ?.sortedBy { it.name }
            ?: emptyList()
        for (file in files) {
            val fileLabel = JLabel("      ${file.name}").apply {
                alignmentX = Component.LEFT_ALIGNMENT
                foreground = PipelineUiHelpers.greenColor
                cursor = java.awt.Cursor.getPredefinedCursor(java.awt.Cursor.HAND_CURSOR)
                addMouseListener(object : java.awt.event.MouseAdapter() {
                    override fun mouseClicked(e: java.awt.event.MouseEvent) {
                        PipelineUiHelpers.openFileInEditor(project, file)
                    }
                })
                PipelineUiHelpers.addHoverEffect(this)
            }
            content.add(fileLabel)
        }
    }

    private fun addChecklists(content: JPanel, paths: FeaturePaths) {
        val checklists = PipelineService.parseChecklists(paths.featureDir ?: "")
        if (checklists.isEmpty()) return

        val totalItems = checklists.sumOf { it.total }
        val completedItems = checklists.sumOf { it.completed }
        content.add(PipelineUiHelpers.sectionHeader("Checklists:  $completedItems/$totalItems complete"))

        for (cl in checklists) {
            val row = JPanel(FlowLayout(FlowLayout.LEFT, 4, 0)).apply {
                isOpaque = false
                alignmentX = Component.LEFT_ALIGNMENT
            }
            val cb = javax.swing.JCheckBox().apply {
                isSelected = cl.isComplete
                isEnabled = false
            }
            row.add(cb)
            val label = JLabel("${cl.fileName}  (${cl.completed}/${cl.total})").apply {
                foreground = if (cl.isComplete) PipelineUiHelpers.greenColor else PipelineUiHelpers.orangeColor
                cursor = java.awt.Cursor.getPredefinedCursor(java.awt.Cursor.HAND_CURSOR)
                addMouseListener(object : java.awt.event.MouseAdapter() {
                    override fun mouseClicked(e: java.awt.event.MouseEvent) {
                        PipelineUiHelpers.openFileInEditor(project, cl.file)
                    }
                })
                PipelineUiHelpers.addHoverEffect(this)
            }
            row.add(label)
            content.add(row)
        }
        content.add(PipelineUiHelpers.verticalSpacer(8))
    }
}

File: speckit-chat-vscode/src/ui/webview/step-detail-panel.ts
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import {
    PipelineStepDef,
    PipelineStepState,
    FeatureEntry,
    StepStatus,
} from '../../model/pipeline-models';
import { pipelineSteps } from '../../service/pipeline-step-registry';
import * as pipelineService from '../../service/pipeline-service';
import * as tasksParser from '../../tools/tasks-parser';
import { getWebviewHtml, escapeHtml, statusBadgeClass } from './webview-utils';

export class StepDetailViewProvider implements vscode.WebviewViewProvider {
    public static readonly viewType = 'speckit.stepDetail';
    private _view?: vscode.WebviewView;
    private _context: vscode.ExtensionContext;

    constructor(context: vscode.ExtensionContext) {
        this._context = context;
    }

    resolveWebviewView(webviewView: vscode.WebviewView): void {
        this._view = webviewView;
        webviewView.webview.options = { enableScripts: true };

        webviewView.webview.onDidReceiveMessage(msg => {
            if (msg.command === 'runStep') {
                vscode.commands.executeCommand('speckit.runStep', msg.stepId, msg.args);
            } else if (msg.command === 'openFile') {
                const uri = vscode.Uri.file(msg.filePath);
                vscode.commands.executeCommand('vscode.open', uri);
            } else if (msg.command === 'saveArgs') {
                this._context.workspaceState.update(`speckit.args.${msg.featureDir}.${msg.stepId}`, msg.args);
            }
        });

        this.showEmpty();
    }

    showStep(step: PipelineStepDef, state: PipelineStepState, featureEntry: FeatureEntry): void {
        if (!this._view) { return; }

        const basePath = vscode.workspace.workspaceFolders?.[0]?.uri.fsPath ?? '';
        const featureDir = featureEntry.path;
        const savedArgs = this._context.workspaceState.get<string>(`speckit.args.${featureEntry.dirName}.${step.id}`) ?? step.defaultArgs;
        const statusLabel = statusText(state.status);
        const badgeClass = statusBadgeClass(state.status);
        const displayNum = displayNumber(step);
        const isReadOnly = step.id === 'specify';

        let body = '';
        body += `<h1>Step ${escapeHtml(displayNum)}: ${escapeHtml(step.name)}</h1>`;
        if (step.isOptional) {
            body += `<p class="muted">(optional)</p>`;
        }
        body += `<p>${escapeHtml(step.description)}</p>`;
        body += `<p><span class="badge ${badgeClass}">${escapeHtml(statusLabel)}</span></p>`;

        // Clarification markers
        if (step.id === 'clarify') {
            const specFile = pipelineService.resolveArtifactFile('spec.md', false, basePath, featureDir);
            if (specFile && fs.existsSync(specFile)) {
                const specContent = fs.readFileSync(specFile, 'utf-8');
                const markerCount = (specContent.match(/\[NEEDS CLARIFICATION/g) || []).length;
                if (markerCount === 0) {
                    body += `<p class="check-pass">No clarification markers in spec.md</p>`;
                } else {
                    body += `<p class="check-fail">${markerCount} [NEEDS CLARIFICATION] marker(s) in spec.md</p>`;
                }
            }
        }

        // Prerequisites
        if (step.prerequisites.length > 0) {
            body += `<div class="section"><h3>Prerequisites:</h3>`;
            for (const result of state.prerequisiteResults) {
                const icon = result.exists ? '\u2713' : '\u2717';
                const cls = result.exists ? 'check-pass' : 'check-fail';
                const detail = result.detail ? ` (${escapeHtml(result.detail)})` : '';
                const fileLink = result.exists && result.resolvedFile && !result.artifact.isDirectory
                    ? ` <a href="#" onclick="openFile('${escapeHtml(result.resolvedFile.replace(/\\/g, '\\\\').replace(/'/g, "\\'"))}')">[open]</a>`
                    : '';
                body += `<div class="check-item"><span class="${cls}">${icon}</span> ${escapeHtml(result.artifact.label)}${detail}${fileLink}</div>`;
            }
            body += `</div>`;
        }

        // Outputs
        if (step.outputs.length > 0) {
            body += `<div class="section"><h3>Outputs:</h3>`;
            for (const result of state.outputResults) {
                const icon = result.exists ? '\u2713' : '\u2717';
                const cls = result.exists ? 'check-pass' : 'check-fail';
                const detail = result.detail ? ` (${escapeHtml(result.detail)})` : '';
                const fileLink = result.exists && result.resolvedFile && !result.artifact.isDirectory
                    ? ` <a href="#" onclick="openFile('${escapeHtml(result.resolvedFile.replace(/\\/g, '\\\\').replace(/'/g, "\\'"))}')">[open]</a>`
                    : '';
                body += `<div class="check-item"><span class="${cls}">${icon}</span> ${escapeHtml(result.artifact.label)}${detail}${fileLink}</div>`;

                // Checklist directory files
                if (step.id === 'checklist' && result.artifact.isDirectory && result.exists && result.resolvedFile) {
                    try {
                        const files = fs.readdirSync(result.resolvedFile)
                            .filter(n => n.endsWith('.md'))
                            .sort();
                        for (const f of files) {
                            const fp = path.join(result.resolvedFile, f);
                            body += `<div class="check-item" style="padding-left: 20px;"><span class="check-pass">\u2713</span> <a href="#" onclick="openFile('${escapeHtml(fp.replace(/\\/g, '\\\\').replace(/'/g, "\\'"))}')">${escapeHtml(f)}</a></div>`;
                        }
                    } catch { /* ignore */ }
                }
            }
            body += `</div>`;
        }

        // Checklists (implement step)
        if (step.id === 'implement' && featureDir) {
            const checklists = pipelineService.parseChecklists(featureDir);
            if (checklists.length > 0) {
                const totalItems = checklists.reduce((s, c) => s + c.total, 0);
                const completedItems = checklists.reduce((s, c) => s + c.completed, 0);
                body += `<div class="section"><h3>Checklists: ${completedItems}/${totalItems} complete</h3>`;
                for (const cl of checklists) {
                    const checkIcon = cl.isComplete ? '\u2713' : '\u25CB';
                    const cls = cl.isComplete ? 'check-pass' : 'check-fail';
                    body += `<div class="check-item"><span class="${cls}">${checkIcon}</span> <a href="#" onclick="openFile('${escapeHtml(cl.filePath.replace(/\\/g, '\\\\').replace(/'/g, "\\'"))}')">${escapeHtml(cl.fileName)}</a> (${cl.completed}/${cl.total})</div>`;
                }
                body += `</div>`;
            }
        }

        // Task list
        if (['tasks', 'implement', 'taskstoissues'].includes(step.id) && featureDir) {
            const tasksFilePath = path.join(featureDir, 'tasks.md');
            const parsed = tasksParser.parse(tasksFilePath);
            if (parsed) {
                body += `<div class="section"><h3>Tasks: ${parsed.completedTasks}/${parsed.totalTasks} complete</h3>`;
                for (const phase of parsed.phases) {
                    const phaseCompleted = phase.tasks.filter(t => t.checked).length;
                    const mvpTag = phase.isMvp ? ' MVP' : '';
                    const priorityTag = phase.priority ? ` (${phase.priority})` : '';
                    body += `<div class="phase-header">Phase ${phase.number}: ${escapeHtml(phase.name)}${escapeHtml(priorityTag)}${escapeHtml(mvpTag)} (${phaseCompleted}/${phase.tasks.length})</div>`;
                    for (const task of phase.tasks) {
                        const checkIcon = task.checked ? '\u2611' : '\u2610';
                        const cls = task.checked ? 'check-pass' : '';
                        const tags = [
                            task.parallel ? '[P]' : '',
                            task.story ? `[${task.story}]` : '',
                        ].filter(Boolean).join(' ');
                        body += `<div class="task-item"><span class="${cls}">${checkIcon}</span><span class="task-id">${escapeHtml(task.id)}</span>`;
                        if (tags) { body += `<span class="task-tags">${escapeHtml(tags)}</span>`; }
                        body += `<span>${escapeHtml(task.description)}</span></div>`;
                    }
                }
                body += `</div>`;
            }
        }

        // Hands off to
        if (step.handsOffTo.length > 0) {
            body += `<p class="muted">Hands off to \u2192 ${escapeHtml(step.handsOffTo.join(', '))}</p>`;
        }

        // Arguments
        body += `<div class="section">`;
        body += `<h3>Arguments:</h3>`;
        body += `<textarea id="argsField" ${isReadOnly ? 'readonly' : ''} rows="3">${escapeHtml(savedArgs)}</textarea>`;
        body += `</div>`;

        // Run button (hidden for Specify)
        if (!isReadOnly) {
            body += `<button onclick="runStep()">Run ${escapeHtml(step.name)} \u25B7</button>`;
        }

        const script = `
            const vscode = acquireVsCodeApi();
            function runStep() {
                const args = document.getElementById('argsField').value.trim();
                vscode.postMessage({ command: 'runStep', stepId: '${step.id}', args });
                vscode.postMessage({ command: 'saveArgs', featureDir: '${escapeHtml(featureEntry.dirName)}', stepId: '${step.id}', args });
            }
            function openFile(filePath) {
                vscode.postMessage({ command: 'openFile', filePath });
            }
        `;

        this._view.webview.html = getWebviewHtml(this._view.webview, `Step: ${step.name}`, body, script);
    }

    showFeature(entry: FeatureEntry, states: Map<PipelineStepDef, PipelineStepState> | undefined): void {
        if (!this._view) { return; }

        let body = '';
        body += `<h1>${escapeHtml(entry.dirName)}</h1>`;
        body += `<p>${entry.completedSteps}/${entry.totalOutputSteps} output steps completed</p>`;

        if (states) {
            body += `<div class="section"><h3>Steps:</h3>`;
            for (const step of pipelineSteps) {
                const state = states.get(step);
                const status = state?.status ?? StepStatus.NOT_STARTED;
                const icon = statusIcon(status);
                const cls = status === StepStatus.COMPLETED ? 'check-pass' :
                    status === StepStatus.BLOCKED ? 'check-fail' : '';
                body += `<div class="check-item"><span class="${cls}">${icon}</span> ${escapeHtml(displayNumber(step))}. ${escapeHtml(step.name)} \u2014 ${escapeHtml(statusText(status))}</div>`;
            }
            body += `</div>`;
        }

        this._view.webview.html = getWebviewHtml(this._view.webview, entry.dirName, body);
    }

    showEmpty(): void {
        if (!this._view) { return; }
        this._view.webview.html = getWebviewHtml(this._view.webview, 'Step Detail', '<p class="muted">Select a pipeline step to view details.</p>');
    }
}

function displayNumber(step: PipelineStepDef): string {
    if (!step.parentId) { return step.number.toString(); }
    const parent = pipelineSteps.find(s => s.id === step.parentId);
    if (!parent) { return step.number.toString(); }
    const siblingIndex = pipelineSteps
        .filter(s => s.parentId === step.parentId)
        .indexOf(step) + 1;
    return `${parent.number}.${siblingIndex}`;
}

function statusIcon(status: StepStatus): string {
    switch (status) {
        case StepStatus.COMPLETED: return '\u2713';
        case StepStatus.READY: return '\u25CB';
        case StepStatus.IN_PROGRESS: return '\u25D0';
        case StepStatus.BLOCKED: return '\u2717';
        default: return '\u25CB';
    }
}

function statusText(status: StepStatus): string {
    switch (status) {
        case StepStatus.COMPLETED: return 'Completed';
        case StepStatus.READY: return 'Ready';
        case StepStatus.IN_PROGRESS: return 'In Progress';
        case StepStatus.BLOCKED: return 'Blocked';
        default: return 'Not Started';
    }
}
